opencensus/metrics/export/gauge.py

```diff
@@ -0,0 +1,222 @@
+# Copyright 2019, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+import six
+import threading
+
+from opencensus.metrics.export import metric
+from opencensus.metrics.export import metric_descriptor
+from opencensus.metrics.export import point as point_module
+from opencensus.metrics.export import time_series
+from opencensus.metrics.export import value as value_module
+
+
+class GaugePointLong(object):
+    """An instantaneous measurement from a LongGauge.
+
+    A GaugePointLong represents the most recent measurement from a
+    :class:`LongGauge` for a given set of label values.
+    """
+
+    def __init__(self):
+        self.value = 0
+        self._value_lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({})"
+                .format(
+                    type(self).__name__,
+                    self.value
+                ))
+
+    def add(self, val):
```

Comment:
total fly by comment as I watch the metric progress: it strikes me as odd for a gauge to have an `add` method instead of just `set` and seems like it could lead to misuse by users, do you know the rationale behind this?

Is it because the value is threadsafe with the lock private, so a user couldn't do an atomic add without the functionality baked in?

Reply:
I think the motivation here is to support data sources that only expose a delta.

I'm following the [java implementation](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/metrics/LongGauge.java#L140-L146) for now, but we need to add a description of the gauges API (including `add`) to [the spec](https://github.com/census-instrumentation/opencensus-specs).

> Is it because the value is threadsafe with the lock private, so a user couldn't do an atomic add without the functionality baked in?

It's actually the other way around: we wouldn't have to lock in `set` if we didn't want to provide an atomic `add`. That said, there's nothing (besides common sense) preventing the user from acquiring the lock themselves and changing the value.